bulldozer/signals.go

```diff
@@ -21,62 +21,186 @@ import (
 	"strings"
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal interface {
+	Enabled() bool
+	// Determine if the signal has values assigned to it and should be considered when matching
+
+	Matches(context.Context, pull.Context, string) (bool, string, error)
+	// Determine if the signal matches a value in the target pull request
+}
+
+type Labels []string
+type CommentSubstrings []string
+type Comments []string
+type PRBodySubstrings []string
+type Branches []string
+type BranchPatterns []string
+type MaxCommits int
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            `yaml:"labels"`
+	CommentSubstrings `yaml:"comment_substrings"`
+	Comments          `yaml:"comments"`
+	PRBodySubstrings  `yaml:"pr_body_substrings"`
+	Branches          `yaml:"branches"`
+	BranchPatterns    `yaml:"branch_patterns"`
+	MaxCommits        `yaml:"max_commits"`
+}
+
+func (signal *Labels) Enabled() bool {
```

Comment:
Slices function similarly to pointers in that they are a small amount of data that refers to more data elsewhere in memory. There's usually no reason to take a pointer to a slice unless you need to modify the slice.

For that reason, I think you'll want to define all of these methods (along with `Matches`) on the value type rather than on the pointer type. This will also save you a bunch of dereferencing.

Same comments apply to the `MaxCommit` functions too.

Reply:
Ok, sounds good